Update partial admission kep to support ElasticJobs (#12281)
kubernetes-sigs/kueue: keps/420-partial-admission/README.md

Reviewer: I think you might need to update this information as this contradicts new algorithm proposed.

Reviewer: Ideally, this point should have been filled as part of the beta process.

Author: Agree, but PartialAdmission feature is already in beta: https://github.com/kubernetes-sigs/kueue/blob/main/pkg/features/kube_features.go#L472

Reviewer: The proportional algorithm (what the code does today) Is the alternative. It should be listed here with the tradeoff: proportional is fairer but less predictable per-PodSet, tail-first is simpler but order-dependent.

Author: I added proportional algorithm as an alternative. 
>what the code does today

Not really. The proportional algorithm is mentioned in this kep, but at the moment we support only one PodSet:
https://github.com/kubernetes-sigs/kueue/blob/main/apis/kueue/v1beta2/workload_types.go#L590 
https://github.com/kubernetes-sigs/kueue/blob/main/pkg/webhooks/workload_webhook.go#L116

Reviewer 2: Please add a user story explaining motivation for the ongoing work.

Author: Added

Reviewer 2: Update for the semantic of the fields is required in the context of the new feature. Specifically we should explain that .spec.count will correspond now to the Job .spec.parallelism as requested by the user.

Author: Done

Reviewer 2: Please refresh the comments, I think we may already have 0 as minimum, and we have use-cases for that for scale-down to 0

Reviewer 2: ```suggestion
    // count is the number of pods requested by the Job.
```

Reviewer 2: ```suggestion
// minCount is the minimum number of pods acceptable for admission
// in case of partial admission.
//
// If not provided, partial admission for the current PodSet is not
// enabled.
// +optional
```

Reviewer 2: Let's also add description and comment for Workload `.status.admission.admittedPodSets[].count`

Author: I think comment for .status.admission.admittedPodSets[].count is up to date, WDYT?

Kueue will not take any measure to ensure that the parent job respects the assigned quota.

> Partial admission is not allowed for concurrent admission in order to not increase the complexity. The webhook will reject workloads that support partial admission to be created for a cluster queue with concurrent admission policy. 

Reviewer 2: Please also add the note to the Concurrent Admission KEP.

## Proposal

Change the way the flavor assigner works to support decrementing the pods count in order to find a better fit for the current workload.
In case a partial fit is chosen, the jobframework reconciler should provide the admitted pod counts to the parent job before unsuspending it, in a similar fashion as the node selectors are provided. In case the job gets suspended, the original pod counts should be restored in order to allow a potential future admission with its original pod counts.

### User Stories

Kueue issue [420](https://github.com/kubernetes-sigs/kueue/issues/420) provides details on the initial feature details and its applicability for `batch/Job`.

> Autoscaled RayJob is running and scaled up from 1000 to 5000 pods, but only capacity for 2000 more pod is available. With partial admission, the Kueue admits a 2000 out of 4000 pods to the cluster instead of rejecting the scale up.

Reviewer 2: Please add structure to the section like 

Story 1 (partial admission)

/  as before + KubeRay

Story 2 (graceful scale up handling)

// as presented currently

Story 3 (opportunistic scale up)

// write a story if this scenario matters as we discussed on a sync

Story 4 (multi-podset RayJob)

// As a user of multi-podset RayJob .. .

Author: Added

## Design Details

> Jobs with `kueue.x-k8s.io/partial-admission=true` annotation are eligible for partial admission.

Reviewer 2: add the definition of rhe annotation in the API section

Author: I added a section of enabling, seems annotation suits there better, wdyt?

### Workload API

```go
type PodSet struct {
    // .......

    // // count is the number of pods requested by the Job.
    // +kubebuilder:validation:Minimum=0
    Count int32 `json:"count"`

    // minCount is the minimum number of pods acceptable for admission in case of partial admission.
    //
    // If not provided, partial admission for the current PodSet is not
    // enabled.
    // +optional
    MinCount *int32 `json:"minCount,omitempty"`
}

type WorkloadStatus struct {
    // ........

    // admission holds the parameters of the admission of the workload by a
    // ClusterQueue. admission can be set back to null, but its fields cannot be
    // changed once set.
    // +optional
    Admission *Admission `json:"admission,omitempty"`
    
    // ........
}

type Admission struct {
    // .........

    // podSetAssignments hold the admission results for each of the .spec.podSets entries.
    // +listType=map
    // +listMapKey=name
    // +kubebuilder:validation:MaxItems=10
    PodSetAssignments []PodSetAssignment `json:"podSetAssignments"`

> }

Reviewer 2: I think these can be skipped, let's keep only PodSetAssignment definition here

type PodSetAssignment struct {
    // ........

    // count is the number of pods taken into account at admission time.
    // This field will not change in case of quota reclaim.
    // Value could be missing for Workloads created before this field was added,
    // in that case spec.podSets[*].count value will be used.
    //
    // +optional
    // +kubebuilder:validation:Minimum=0
    Count *int32 `json:"count,omitempty"`

> }

Reviewer 2: Please add a section on the "Validation" for the API fields. Some ideas:

minCount <= count
minCount <= admission.count <= count

wdyt?

Author: I think validation for minCount is the Job responsibility, we can just have an partial admission error

Reviewer 2: we probably should also have a validation at the workload object to prevent malicious settings or just incorrect settings. I think the mechanism we do coukd also be used for custom job crd imementations in general.

Finally defining invartiants on the workload crd helps other developers to understand the code

Author: Added validation

```

> ### Scheduler / Flavorassignment

Reviewer 2: On a sync you also described the "opportunistic scale up" - scale up after partial admission when capacity becomes available. We discussed this could be done by the ElasticWorkloads, by sending the new workload probe. 

Please either include design for this in the doc as a dedicated section, or explicitly make out of scope.

Author: added the section in design details of elasticjobs

#### Partial Admission for one PodSets

The search for appopriate count value is done using binary search algorithm.

#### Partial Admission for multiple PodSets

The are multiple ways how to approach multiple podsets shrinking in case of insufficient quota. For simplisity reason we'll start with the order-based one and will expand option if needed in future.
Another approch that was considered is proportional. However it was disgarded because of insufficiency for some use cases.

- **`order-based` (default)**: Shrinks the counts of the PodSets sequentially starting from the last one (suits for the cases when the podsets are ordered by priority). The Workload PodSet order is usually the same as the order of the PodSet in the Job spec. For the RayCluster the Workload PodSets starting from Head PodSet and following by WorkerGroupPodSets in the same order as in RayCluster.

#### Order-Based policy (`order-based`)

> Under the `order-based` policy, Kueue shrinks the PodSets starting from the last one and moving towards the beginning as needed.

Reviewer 2: what determines the order of PodSets being considered in this strategy? Names or declaration order in the array list?

Author: declaration order in the array list

| **batch/Job** | `true` | `true` | **Partial Admission + Workload Slicing**: Kueue can shrink the Job's podset down to its `job-min-parallelism`. The Job's spec (`.spec.parallelism`) remains unmodified. Kueue uses scheduling gates to ungate only the admitted pod count. A pending workload slice is created to dynamically scale up (opportunistic scale up) when more quota becomes available. |
| **batch/Job** | `true` | `false`/not present | **Partial Admission (Traditional)**: Kueue shrinks the Job's podset down to its `job-min-parallelism`. Upon admission, Kueue modifies the Job's spec (`.spec.parallelism`) directly to match the admitted count. No scheduling gates are used, and no future scale-up occurs for this workload. |
| **batch/Job** | `false`/not present | `true` | **Workload Slicing Only**: Kueue cannot shrink the Job's podset during initial scheduling; the job must fit its full requested capacity to be admitted. If admitted, scheduling gates are used. If the user manually scales up the Job, Kueue creates a new workload slice for the scale-up request, but this scale-up will only be admitted if it can be fully satisfied (no partial admission of the scale-up). |

> | **KubeRay** | `true` | `true` | **Partial Admission + Workload Slicing**: Kueue can shrink the worker groups down to their `minReplicas` (defined in the worker spec). The Ray object's spec remains unmodified. Admission is managed via scheduling gates. A pending workload slice is created to allow KubeRay to scale up (opportunistic scale up) worker pods as more quota becomes available. |

Reviewer 3: Does it affect head pods in any way?

Author: Head pod doesn't have minReplicas, so no

#### Story 3 (opportunistic scale up)

As a user running an elastic job (like a RayJob), when my job is admitted with partial capacity due to resource constraints, I want the job to dynamically scale up to its full requested capacity as soon as other workloads complete and resources become available in the cluster, maximizing resource utilization and reducing the job's overall completion time.

#### Story 4 (multi-podset RayJob)

As a user of a multi-podset RayJob (which defines a head pod and multiple worker groups, potentially targeting different resource flavors or node groups like reservation, on-demand, or spot), I want to enable partial admission such that Kueue reduces the worker groups sequentially starting from the least critical (e.g., spot or low-priority worker groups defined last in the spec) while preserving the capacity of the more critical worker groups.

> ## Design Details

Reviewer 2: I think there is also a problem (even currently) with ElasticJobs for KubeRay, but for partial admission in particular. 

Because PodsReady requires all pods to be ready, but we should only expect that from the number of admitted pods, ie Workloads `.status.admission.podSets.count` - I think we should actually open an issue for this even today. Would you like to drive fixing this since this is needed for KubeRay elastic jobs anyway?

Author: ```
func (j *RayCluster) PodsReady(ctx context.Context, _ client.Client) bool {
	return j.Status.State == rayv1.Ready
}
```

The implementation doesn't rely on podSets.count

Reviewer 2: Can you investigate if this is correct for partial admission?  I'm wondering if partially admitted RayCluster will have Ready=True. If this is the case, then I suppose it is ok.

Author: >I'm wondering if partially admitted RayCluster will have Ready=True.

RayCluster will have Ready state only if all pods are running. However the state will remain Ready when the RayCluster is scaled up but not all newly created pods are running. 

So I suggest to add it as limitation for now. I.e. when the job is created, it could not be partially admitted, the kueue will partially admit only scale up. We can also recommend to user to set count=minCount, because the main user request was regarding scale ups.

Author: For RayCluster when constructing the initial (first) the minCount will equal replicas and then minCount for following Workloads (scaleups) will equal to admitted count + 1.

Reviewer 4: I did a preliminary test on a RayCluster:
1. During a scale-up, the RayCluster is ready even if the new pods are not ready. I did this by bumping `replicas` for the cluster and adding a scheduling gate to the new pod.
2. During initial admission, if not all workers are ready, the RayCluster will not be ready (and it will have the `RayClusterProvisioned: False` Condition). I did this by using the `podAntiAffinity` on the workers, so one of them gets unschedulable during init.

I'll confirm both further, but what 2. suggests to me is that partial admission might be broken. Because an initial config that explicitly sets `replicas` to something greater than `minReplicas` is a valid configuration, no?

So in the case where we set `minReplicas: 1, replicas: 4`, Kueue will set `MinCount=1` and might do a partial admission of 1-3 workers. But according to finding 2. this RayCluster will not become ready until the 4th pod is ready.

> We can also recommend to user to set count=minCount

Hm, but wouldn't this mean that there really is no partial admission? Because if `count=minCount`, then this is just regular admission for RayCluster, no? We can't admit it partially, we can only scale it up.

Author: > We can't admit it partially, we can only scale it up.

True. However the main user request was regarding partial scale up. So I'm thinking now, that we should not mix partial scale up and partial admission despite both are using the one mechanism.

Reviewer 4: > However the main user request was regarding partial scale up.

Which to me sounds more like a feature for Elastic Jobs, a second mode of operation to differentiate the all-or-nothing scale up that is implemented currently.

In general, to me it seems that partial admission could be treated under the Elastic Job feature, conceptually. But that is not a conversation for right now.

Author: I created a separate kep for partial scaleup: https://github.com/kubernetes-sigs/kueue/pull/14310

Reviewer 2: I would prefer to consolidate this all in one PR. Seems like the topics are too inter-connected to spread in multiple PRs imo.

Author: Sure. I'm closing this PR

### Enablement

PartialAdmission in Kueue is enabled through a combination of a Kubernetes feature gate and an opt-in annotation on individual Workload objects. At the cluster level, the PartialAdmission feature (enabled by default) and PartialAdmissionForElasticJobs feature for ElasticJobs, must be enabled via the corresponding Kueue feature gate.

Once the feature gate is enabled, individual Job objects can opt into partial admission by including the ``kueue.x-k8s.io/partial-admission="true"` annotation. 
When both conditions are met, Kueue treats the Workload as eligible for partial admission. 

#### Features
```go
	// Enables partial admission for non-elastic jobs.
	PartialAdmission featuregate.Feature = "PartialAdmission"

	// Enables partial admission for elastic jobs.
	PartialAdmissionForElasticJobs featuregate.Feature = "PartialAdmissionForElasticJobs"
```

#### Partial Admission Annotation
```go
const (
  // EnabledAnnotationKey refers to the annotation key present on Job's that support
  // workload slicing.
  // This annotation is alpha-level.
  EnabledPartialAdmission = "kueue.x-k8s.io/partial-admission"
)
```

#### Scenario Summary

> The table below summarizes the scheduling and admission behaviors for different combinations of the `kueue.x-k8s.io/partial-admission` and `kueue.x-k8s.io/elastic-job` annotations for `batch/v1.Job` and KubeRay (`RayJob` / `RayCluster`) objects:

Reviewer 2: Thank you for the table, this clarifies a lot.

| **KubeRay** | `true` | `true` | **Partial Admission + Workload Slicing**: Kueue can shrink the worker groups down to their `minReplicas` (defined in the worker spec). The Ray object's spec remains unmodified. Admission is managed via scheduling gates. A pending workload slice is created to allow KubeRay to scale up (opportunistic scale up) worker pods as more quota becomes available. |
| **KubeRay** | `true` | `false`/not present | **Not supported / Out of scope**: Kueue does not support partial admission for non-autoscaled KubeRay jobs at this time. |
| **KubeRay** | `false`/not present | `true` | **Workload Slicing Only**: Kueue cannot shrink the worker groups; they must fit their full requested replicas to be admitted. Once admitted, scheduling gates are used. If the autoscaler scales up the worker groups, Kueue creates a workload slice representing the scale-up request, but this scale-up is only admitted if there is enough quota to satisfy the full scale-up request. |

### Workload API

```go

> // +kubebuilder:validation:XValidation:rule="has(self.minCount) ? self.minCount <= self.count : true", message="minCount should be positive and less or equal to count"

Reviewer 2: Insteead of the kubebuilder, please define a section with the validation constraints. I think this is actually more involving that that.

We don't need to have a rule which compiples exactly, what is more important to me is general direction:
- `.spec.podSets.minCount <= .spec.podSets.count`
- `.spec.podSets.minCount <= .status.admission.podSets.count <= .spec.podSets.count`
- decide if minCount= 0 is supported (I think so, but we may state that explicitly)

You may also mention that the exact arythmentics is impacted by the reclamablePods.

cc @amy who is now an expect in reclamablePods :)

Author: Added validation for `.spec.podSets.minCount >= 0` and `.spec.podSets.minCount <= .spec.podSets.count.`
I think this validation should happen upon creation. 

For `.spec.podSets.minCount <= .status.admission.podSets.count <= .spec.podSets.count` i don't see when we should validate it. This is a part of scheduling

type PodSet struct {
  // .......

  // // count is the number of pods requested by the Job.
  // +kubebuilder:validation:Minimum=0
  Count int32 `json:"count"`

  // minCount is the minimum number of pods acceptable for admission in case of partial admission.
  //
  // If not provided, partial admission for the current PodSet is not
  // enabled.
  // +optional
  MinCount *int32 `json:"minCount,omitempty"`
}

type PodSetAssignment struct {
  // ........

  // count is the number of pods taken into account at admission time.
  // This field will not change in case of quota reclaim.
  // Value could be missing for Workloads created before this field was added,
  // in that case spec.podSets[*].count value will be used.
  //
  // +optional
  // +kubebuilder:validation:Minimum=0
  Count *int32 `json:"count,omitempty"`
}
```

### Scheduler / Flavorassignment

In case the workload proposed for the current scheduling cycle does not fit, with or without preemption, in the current available quota and any of its PodSets allow partial admission, kueue will try to find a lower counts combination that fits the available quota with or without borrowing.

> #### Partial Admission for one PodSets

Reviewer 2: "one PodSets" - something is wrong here

> The search for appopriate count value is done using binary search algorithm.

Reviewer 2: appopriate - is a typo, please check the KEP with spellchecker

#### Partial Admission for multiple PodSets

> The are multiple ways how to approach multiple podsets shrinking in case of insufficient quota. For simplisity reason we'll start with the order-based one and will expand option if needed in future.

Reviewer 2: simplisity is a typo

> Another approch that was considered is proportional. However it was disgarded because of insufficiency for some use cases.

Reviewer 2: Move the mention of this to alternatives section

> Another approch that was considered is proportional. However it was disgarded because of insufficiency for some use cases.

Reviewer 2: "disgarded" - something is wrong, please check with spellcker

### Scheduler / Flavorassignment

In case the workload proposed for the current scheduling cycle does not fit, with or without preemption, in the current available quota and any of its PodSets allow partial admission, try to find a lower counts combination that fits the available quota with or without borrowing.

> The search could be priority-based or proportionally. It's a job-level setting, that is set by adding 'kueue.x-k8s.io/partial-admission-policy=priority-based' or 'kueue.x-k8s.io/partial-admission-policy=proportional' annotation. If the value is not set, the proportional policy is used. 

Reviewer 5: Is this a new annotation or has it been defined elsewhere? https://kueue.sigs.k8s.io/docs/reference/labels-and-annotations doesn't mention it.

Author: This is a new annotation.

> The priority-based policy assumes that the PodSets are listed in descending order of priorities.

Reviewer 5: Can we get more description on the policies? Ideally with some examples.

Author: Done

The accepted number of pods in each PodSet is recorded in `workload.Status.Admission.PodSetAssignments[*].ResourceUsage.Count`

In order to evaluate the potential success of the preemption, the preemption process should be split into:
- Target selection, which should select the workloads within cohort that will need to be evicted, if the list is empty the assignment will be treated as `NoFit`. This step will take place during nomination.
- Preemption issuing, will do the actual eviction of the targets previously selected.

**Partial admission with multiple variable PodSets count**

> If multiple PodSets within a workload have variable counts, Kueue will iterate over the PodSets in the order they are defined in the Workload spec, starting from the end, and will try to decrease their counts until a fit is found. If shrinking the last PodSet still results in a NoFit, Kueue will try to decrease the next PodSet while keeping the last one at its minimum count, and so on.

Reviewer 5: Do we need to support multiple partial admission pod-set? What is the use case?

Author: Supporting partial admission across only a single PodSet is highly limiting. 
Current partial admission in Kueue relies on the kueue.x-k8s.io/job-min-parallelism annotation. This annotation only allows you to define a single minimum count for the entire job.

The primary use case is support for elastic workloads like an autoscaling RayCluster, which inherently split their work into multiple distinct PodSets (such as a Head node group and one or more elastic Worker node groups).
For example, if a RayCluster needs 1 Head pod (which is non-negotiable) and a variable number of Worker pods (e.g., minimum 5, target 20), a single global parameter (kueue.x-k8s.io/job-min-parallelism annotation that kueue currently support) is not flexible enough to negotiate different thresholds for different pod sets.

In case the workload proposed for the current scheduling cycle, does not fit, with or without preemption, in the current available quota and any of its PodSets allow partial admission, try to find to find a lower counts combination that fits the available quota with or without borrowing.

> The search should be priority based assuming that PodSets are listed accordinly to priorities.

Reviewer: ```suggestion
The search should be priority based assuming that PodSets are listed accordingly to priorities.
```

### Scheduler / Flavorassignment

In case the workload proposed for the current scheduling cycle does not fit, with or without preemption, in the current available quota and any of its PodSets allow partial admission, try to find a lower counts combination that fits the available quota with or without borrowing.

The policy for shrinking PodSet counts is configured via the `kueue.x-k8s.io/partial-admission-policy` annotation on the Job. The supported policies are:
- **`proportional` (default)**: Shrinks the counts of all variable PodSets proportionally to their allowable reduction ranges.

> - **`priority-based`**: Shrinks the counts of the PodSets sequentially, assuming they are listed in descending order of priority (i.e., the first PodSet has the highest priority and the last PodSet has the lowest priority).

Reviewer 5: Explain more clearly how the list is provided - podset importance/priority matches the order they are mentioned in the workload object. Mention how the the order in the podset matches the order in jobset/rayjob/lws/whatever.

Author: Updated

> - **`priority-based`**: Shrinks the counts of the PodSets sequentially, assuming they are listed in descending order of priority (i.e., the first PodSet has the highest priority and the last PodSet has the lowest priority).

Reviewer 5: What about renaming it to something that doesn't have priority? "Priority" is already booked for workload/pod numerical priority and may cause confusion. Maybe something like `podset-order-based`?

Author: Done

#### Priority-Based Policy (`priority-based`)

> Under the `priority-based` policy, Kueue assumes that the PodSets are listed in descending order of priority. To preserve the counts of higher-priority PodSets, Kueue shrinks the PodSets starting from the end (lowest priority) and moving towards the beginning (highest priority) as needed.

Reviewer 5: Can you mention the real world situation (not just an theoretical example) where this strategy is more suitable than proportional?

Author: Added below

> Specifically, if multiple PodSets have variable counts, Kueue iterates over them in the order they are defined in the Workload spec, starting from the last one. It decreases the count of the current PodSet down to its `minCount` until the workload fits the available quota. If shrinking the last PodSet to its `minCount` is still not enough to fit, Kueue keeps it at its `minCount` and moves to the second-to-last PodSet, decreasing its count down to its `minCount`, and so on.

Reviewer 5: We have a similar problem in preemption algorithm. First we try to find a set of preemption candidates that allows to fit the considered workload at all, then we gradually try to put the unneeded candidates back, to ensure that the preemption list is somewhat minimal.

Here, if I understand correctly, you try to find a sizing that fits, but you don't try to maximize the number of pods once you find a workable sizing (no phase 2 equivalen of the preemption algorithm). Is this desired?

Author: Added a phase 2 to the partial admission as well.

  2. Since it does not fit the quota of 10, Kueue keeps `ps2` at `10` and moves to the next lowest priority PodSet, `ps1`.
  3. Kueue decreases `ps1` to its minimum: `2` (reduction of 2 pods). The current total count is now `1 + 2 + 10 = 13`.
  4. Since it still does not fit the quota of 10, and the remaining PodSet `ps0` does not allow partial admission (has no `minCount`), the search fails.
  5. The job remains unadmitted.

#### Proportional Policy (`proportional` - default)

Under the `proportional` policy, Kueue reduces the counts of all variable PodSets proportionally to their maximum allowable reduction range (i.e., `count - minCount`).

For each PodSet $j$ that allows partial admission, let:
- $C_j$ be the maximum requested count.
- $M_j$ be the minimum acceptable count (`minCount`).
- $D_j = C_j - M_j$ be the maximum reduction delta for PodSet $j$.
- $T_D = \sum D_j$ be the sum of all maximum reduction deltas.

To find the optimal fit, Kueue performs a binary search on a scaling index $i$ in the range $[0, T_D]$. For a given index $i$, the count for PodSet $j$ is computed as:

> $$currentCount_j = C_j - \lfloor \frac{D_j \times i}{T_D} \rfloor$$

Reviewer 5: How the pod count is rounded?

Author: To the lower or equal int: https://github.com/kubernetes-sigs/kueue/blob/main/pkg/scheduler/flavorassigner/podset_reducer.go#L59

- **`proportional` (default)**: Shrinks the counts of all variable PodSets proportionally to their allowable reduction ranges.
- **`order-based`**: Shrinks the counts of the PodSets sequentially starting from the last one (suits for the cases when the podsets are ordered by priority). The Workload PodSet order is usually the same as the order of the PodSet in the Job spec. For the RayCluster the Workload PodSets starting from Head PodSet and following by WorkerGroupPodSets in the same order as in RayCluster.

#### Order-Based Policy (`order-based`)

Under the `order-based` policy, Kueue shrinks the PodSets starting from the last one and moving towards the beginning as needed.
Specifically, if multiple PodSets have variable counts, Kueue iterates over them in the order they are defined in the Workload spec, starting from the last one. It decreases the count of the current PodSet down to its `minCount` until the workload fits the available quota. If shrinking the last PodSet to its `minCount` is still not enough to fit, Kueue keeps it at its `minCount` and moves to the second-to-last PodSet, decreasing its count down to its `minCount`, and so on.

> As an optimization, we will introduce a second phase (similar to the preemption algorithm): when a workload finds a combination that fits the available quota, Kueue tries to gradually put the reduced counts back. In this phase, Kueue iterates over all PodSets from the first to the last one. For each PodSet that was reduced, Kueue tries to increase its count back to the original count. If that fits, Kueue keeps it. Otherwise, Kueue performs a binary search on the PodSet's count between the current count and the original count to find the maximum count that fits.

Reviewer 5: Can we have an example covering phase 2?

Author: Added Scenario D to cover it

### Scheduler / Flavorassignment

In case the workload proposed for the current scheduling cycle does not fit, with or without preemption, in the current available quota and any of its PodSets allow partial admission, try to find a lower counts combination that fits the available quota with or without borrowing.

> The policy for shrinking PodSet counts is configured via the `kueue.x-k8s.io/partial-admission-policy` annotation on the Job. The supported policies are:

Reviewer 2: Please define the annotation in the Design Details section for the API.

Author: I added it above API section

> - **`proportional` (default)**: Shrinks the counts of all variable PodSets proportionally to their allowable reduction ranges.

Reviewer 2: Actually, I think the proportional has signifficant issues:
1.  when handling heterogenous PodSets - it would scale both CPU and GPU.
2. even for homogenous PodSets it cannot find simple solutions like (1,2), or (2, 1) to scale down from (2,2). It will just find (1,1) [assuming we have quota for 3].

So, I would like to request parking the proportional mode. Unless we can expand the User story section with a use case which really requires this policy

Reviewer 2: Also, the good part about dropping proportional is that we would not need to introduce the `kueue.x-k8s.io/partial-admission-policy` annotation at all for now.

Author: I verified with the user, that they use only 1 working group, so they don't need a specific policy at the moment — anything will work for them. 

So, I agree to leave only order-based algorithm. Removed proportional algorithm.

> - **`order-based`**: Shrinks the counts of the PodSets sequentially starting from the last one (suits for the cases when the podsets are ordered by priority). The Workload PodSet order is usually the same as the order of the PodSet in the Job spec. For the RayCluster the Workload PodSets starting from Head PodSet and following by WorkerGroupPodSets in the same order as in RayCluster.

Reviewer 2: Should we name it `greedy-reduce-with-fillback` to better reflect on the mechanism .The fact this is order-based does not really tell much by itself.

Author: I though the order is defined by user, so the order-based should be a good name.

> #### Order-Based Policy (`order-based`)

Reviewer 2: Note that currently PodSets in our API are not ordered. So this policy needs to first define the order.
One idea is to use alphabetical order, but it should be explicitly noted.

Author: I thought the order is the order of the PodSets in the Job. 
For the RayCluster/RayJob/RayService the PodSets order in Workload is HeadPod, Submitter jobs etc, and then WorkingGroups in the same order as they are in the Job. We are interested only in WorkingGroups, because only they could have min size.

**Partial admission with multiple variable PodSets count**

If multiple PodSets within a workload have variable counts, Kueue will iterate over the PodSets in the order they are defined in the Workload spec, starting from the end, and will try to decrease their counts until a fit is found. If shrinking the last PodSet still results in a NoFit, Kueue will try to decrease the next PodSet while keeping the last one at its minimum count, and so on.

> As an optimisation we will introdce an increased step: when a workload finds a combination that fits the available quota, Kueue could try to find the combination that can be reached with minimum pod loss by checking if any of the PodSets with count == min count can be increased up to the original count. 

Reviewer: ```suggestion
As an optimization we will introduce an increased step: when a workload finds a combination that fits the available quota, Kueue could try to find the combination that can be reached with minimum pod loss by checking if any of the PodSets with count == min count can be increased up to the original count. 
```

> The increased step is not nessesary for elastic workloads, but it the non-elastic workloads will benefit from it.

Reviewer: ```suggestion
The increased step is not necessary for elastic workloads, but the non-elastic workloads will benefit from it.
```

**Partial admission with multiple variable PodSets count**

> If multiple PodSets within a workload have variable counts, Kueue will iterate over the PodSets in the order they are defined in the Workload spec, starting from the end, and will try to decrease their counts until a fit is found. If shrinking the last PodSet still results in a NoFit, Kueue will try to decrease the next PodSet while keeping the last one at its minimum count, and so on.

Reviewer: The change to tail-first reduction makes sense for Ray-style workloads where later PodSets (e.g., GPU workers) are more expendable than earlier ones (e.g., CPU preprocessors), but worth stating explicitly that podSets ordering acts as reduction priority.

Reviewer: Can you also cover the recovery step in more detail? It would help to include a worked example where recovery is actually triggered, as the current example doesn't demonstrate that path.

Author: What do you mean by recovery?

Reviewer: By "recovery" I meant the increase step described in the KEP after shrinking from the tail to find a fit (e.g., reaching {1, 2, 10} = 13 pods), showing how Kueue tries increasing each PodSet back toward its original count to reach the minimum-loss combination ({1, 4, 14} = 19 pods). The current example only shows the final result, not that intermediate step.

Author: In the current example:
1. The starting point is {1, 4, 20}.
2. Kueue tries to shrink the last podSet (using the existing approach) and finds the {1, 4, 14} combination.
3. If {1, 4, 10} still doesn't fit, Kueue evaluates {1, 2, 10} and {1, 3, 10}.

We could also add an additional step: once a suitable combination is found, Kueue tries to increase the replicas from the last podSet to find the absolute optimal combination (similar to what we do in preemption). However, I didn't include that here because this change specifically targets elastic workloads. For elastic workloads, we will have a WorkloadSlice that represents the original count and will handle verifying the increased replicas for that slice.

Reviewer: Makes sense! the WorkloadSlice recovery path covers this for elastic. Worth noting that non-elastic multi-PodSet partial admission would benefit from the increase step if we support it later, but fine to defer.

Author: Done

And only being able to admit 19 pods, it should end up with

```go
[]podSet { {Count: 1} , {Count: 4}, {Count: 14}}
```

> If the priority based partial admission is not sufficient, the "proportional" partial admission will be introduced.

Reviewer: Can you please elaborate this point? like what triggers the switching?

Author: I meant that we can introduce "proportional" partial admission if we have explicit request for it. Actually not sure I need to mention it here. Moved proportional algorithm to the alternatives section.

Whether an MPIJob supports partial admission or not can be deduced based on `MinAvailable` without the need of a dedicated annotation.
Additional research is needed into the potential usage of multiple variable count PodSets.

### RayJob/RayService/RayCluster controller

> The RayCluster with 'kueue.x-k8s.io/partial-preemption=true' annotation are eligible for partial preemption. The

Reviewer: should this be `partial-admission` to match existing terminology (PartialAdmission FG, partialAssignment struct)?

RayCluster.workerGroupSpec[i].minReplicas will be translated to the PodSet.MinCount for the Worker PodSet.
There will be no update on RayCluster object. If the number of admitted pods less than requested count, the leftover pods should remain scheduling gated. In order to achieve that the same mechanism as for elastic workloads will be applied – the podTemplate will be initially gated and then the correct amount of pods will be ungated.

> ### Test Plan

Reviewer: While we are here, could you also update the test plan?

Author: updated test plan by including the existing tests and adding test for priority based algorithm

Whether an MPIJob supports partial admission or not can be deduced based on `MinAvailable` without the need of a dedicated annotation.
Additional research is needed into the potential usage of multiple variable count PodSets.

### RayJob/RayService/RayCluster controller

The RayCluster with 'kueue.x-k8s.io/partial-preemption=true' annotation are eligible for partial preemption. The
RayCluster.workerGroupSpec[i].minReplicas will be translated to the PodSet.MinCount for the Worker PodSet.

> There will be no update on RayCluster object. If the number of admitted pods less than requested count, the leftover pods should remain scheduling gated. In order to achive that the same mechanism as for elastic workloads will be applied – the podTemplate will be initially gated and then the correct amount of pods will be ungated.

Reviewer: ```suggestion
There will be no update on RayCluster object. If the number of admitted pods less than requested count, the leftover pods should remain scheduling gated. In order to achieve that the same mechanism as for elastic workloads will be applied – the podTemplate will be initially gated and then the correct amount of pods will be ungated.
```

## Implementation History
- 07/07/2023 - Partial admission for batch.job was added that support only 1 podSet with minCount value.

## Drawbacks

## Alternatives

> Using proportional algorithm instead of priority based one, i.e each PodSet should be shrinked proportionaly.

Reviewer: ```suggestion
Using proportional algorithm instead of priority based one, i.e each PodSet should be shrunk proportionaly.
```

## Implementation History
- 07/07/2023 - Partial admission for batch.job was added that support only 1 podSet with minCount value.

> ## Drawbacks

Reviewer: The same here.

Reviewer: https://github.com/kubernetes-sigs/kueue/pull/12281/changes#r3431196655

```

### ElasticJob

For ElasticJobs, updating `job.spec.parallelism` or `job.spec.count` could cause race conditions between partial admission and scaling up/down activity. 
To avoid this, the `podSets` count won't be updated in `RunWithPodSetsInfo` for elastic jobs. Instead, the workload controller will use the `workload.Status.Admission.PodSetAssignments[*].Count` value to calculate the number of pods from which Kueue should remove scheduling gates. 
The `minCount` value for an ElasticJob workload will represent the currently admitted value + 1.
Also, a new Workload representing the full job will be created and added to the queue, to admit the remaining capacity once it becomes available.

#### Example: Two-Step Scale Up under Quota Constraints (with Partial Admission)

Consider a scenario where:
1. The ClusterQueue has a total quota of **7** for the requested resource flavor.
2. The Job is configured for both `ElasticJobs` and `PartialAdmission`.
3. The user performs a two-step scale up of the Job: starting at **5** replicas, scaling up to **10**, and then to **12**.

> ##### Step 0: Job Creation (Initial Size: 5)

Reviewer 2: Instead of doing the one chained step-ed examples, could you show multiple transition cases rather than steps. 

The relevant examples:
- partial admission succeeded, partial
- scale up of the Job by user, succeeding fully
- scale up of the Job by user, succeeding partially
- opportunistic scale up when capacity is freed

Author: I added a name for each step, so it should be linked transition cases. Is that what you meant?

* **Job spec.parallelism**: 5
* **Workloads**:
  * `wl-A` (Admitted):
    * `spec.podSets.count` = 5
    * `spec.podSets.minCount` = 2
    * `status.admission.count` = 5
* **Controller Actions**:
  1. **Job Controller**: Detects the Job creation, sets `.spec.suspend = false`, and creates 5 Pods. Due to Kueue's webhook mutation, the Pods are created with the `kueue.x-k8s.io/elastic-job` scheduling gate.
  2. **Kueue Job Framework / Workload Controller**: Detects the Job and creates `wl-A` with `spec.podSets.count = 5` and `spec.podSets.minCount = 2`.
  3. **Kueue Scheduler**: Evaluates `wl-A`. Since the requested 5 pods fit within the available quota of 7, it admits `wl-A` (`status.admission.count = 5`), reserving 5 units of quota.
  4. **ElasticJobUngater Controller**: Detects that `wl-A` is admitted and removes the scheduling gate from the 5 pods.
  5. **Kube-scheduler**: Schedules the 5 ungated pods, which transition to the Running state.
* **Quota usage**: 5/7 (2 available).

##### Step 1: Scale Up from 5 to 10
* **Job spec.parallelism**: 10
* **Workloads**:
  * `wl-A` (Finished - aggregated/replaced by `wl-B`)
  * `wl-B` (Admitted - Partially):

>     * `spec.podSets.count` = 7 (originally requested 10, but updated to 7 during partial admission)

Reviewer 2: I think this is incorrect given our recent discussion  - the wl .spec.podSet.count stays as in Job, but the

  3. **Kueue Scheduler**: Evaluates `wl-C`. The demand is `12 - 7 = 5`. The available quota is 0, so the scheduler puts `wl-C` in the queue.

##### Step 3: Quota increases to 12
If the available quota in the ClusterQueue increases to 12 (or more) in the future:
* **Workloads**:
  * `wl-B` (Finished)
  * `wl-C` (Admitted):
    * `spec.podSets.count` = 12
    * `spec.podSets.minCount` = 8 (7 + 1)
    * `status.admission.count` = 12

In the next scheduler loop, the Kueue scheduler will evaluate and admit `wl-C`. It will also update its `spec.podSets.minCount` to match the job's minCount.

### batch/Job controller

Besides adapting `RunWithPodSetsInfo` and `RestorePodSetsInfo` it should also:

- rework `PodSets()` to populate `MinCount` if the job is marked to support partial admission.
  * jobs supporting partial admission should have a dedicated annotation, e.g., `kueue.x-k8s.io/job-min-parallelism`, indicating the minimum `parallelism` acceptable by the job in case of partial admission.
  * jobs which need the `completions` count kept in sync with `parallelism` should indicate this in a second annotation, `kueue.x-k8s.io/job-completions-equal-parallelism`

> - rework `EquivalentToWorkload` to account for potential differences in `PodSets` spec `Parallelism`.

Reviewer 2: Please also add a paragraph below that the Job will be migrated to the new mechanism under the PartialAdmissionForElasticJobsForBatchJob FG

Author: >the Job will be migrated to the new mechanism

Actually I don't think any migration is needed. In order to apply elastic approach, the workload should have elastic-job annotation. If the job has elastic-job annotation, during the partial admission the elastic path will be chosen. 
If the workload doesn't have elastic-job annotation, kueue won't apply workload-slices mechanism.

Reviewer 2: Basically my point is that end the end of the day we don't want to maintain two mechanisms for partial admission. 

This needs to be clarified in the KEP. 

what happens on RayJob and Job when user sets:
- partial-admission=true & elastic-job-true
- partial-admission=true only
- elastic-job=true only
I was thinking that you want to make `partial-admission=true` imply `partial-admission=true` & `elastic-job-true`, but I'm no longer sure. 

Please clarify the semantics for the combination of the annotations. My only request is to align the semantics for all CRDs.

Author: I think approaches of partial admission for elastic workloads and non-elastic workloads are different and should not be mixed.

Reviewer 2: Sorry, I'm very confused. I would like to request a summary in the KEP the behavior in the three scenarios as I listed above:
1. `partial-admission=true` & `elastic-job-true`
2. `partial-admission=true` only
3. `elastic-job=true`

Then multiply this by Job and KubeRay. 

As a maintainer of Kueue I would like to understand if we need to maintain separate handling for Job and KubeRay. So far I cannot understand the argument.

Author: Added

Whether an MPIJob supports partial admission or not can be deduced based on `MinAvailable` without the need of a dedicated annotation.
Additional research is needed into the potential usage of multiple variable count PodSets.

### RayJob/RayService/RayCluster controller

> RayClusters with 'kueue.x-k8s.io/partial-admission=true' annotation are eligible for partial admission. The

Reviewer 2: Please add the annotation to the Design Details APi section

RayCluster.workerGroupSpec[i].minReplicas will be translated to the PodSet.MinCount for the Worker PodSet.

> There will be no update on RayCluster object. If the number of admitted pods is less than requested count, the leftover pods should remain scheduling gated. In order to achieve that, the same mechanism as for elastic workloads will be applied – the podTemplate will be initially gated and then the correct amount of pods will be ungated.

Reviewer 2: I think we should actually aim with the mechanism to replace the old one also for Job. Maintaining two different mechanism depending on the Job CRD will be nightmare. 

So I want to propose to introduce a new FG like `PartialAdmissionViaElasticJobsForJob` to allow eventual migration of the pre-existing mechanism to the new one also for Job.

Author: I added a feature gate. 
However I would not mix approach for ElasticJobs and regular jobs for now. When we can see that approach for ElasticJobs works fine, we can extend it to non-elastic job

```

### ElasticJob

> For ElasticJobs, updating `job.spec.parallelism` or `job.spec.count` could cause race conditions between partial admission and scaling up/down activity. 

Reviewer 2: `job.spec.count` - do you mean here Workload's `.spec.podSet[].count`? I'm not sure

To avoid this, the `podSets` count won't be updated in `RunWithPodSetsInfo` for elastic jobs. Instead, the workload controller will use the `workload.Status.Admission.PodSetAssignments[*].Count` value to calculate the number of pods from which Kueue should remove scheduling gates. 

> The `minCount` value for an ElasticJob workload will represent the currently admitted value + 1.

Reviewer 2: I'm confused here, two questions:
1. by `minCount` do you mean Workload's `.spec.podSets[].minCount`? If so, please clarify
2. `minCount` is `admitted value+1`? 

This sounds counter-intuitive. I would assume `.spec.podSets[].count` `.spec.podSets[].minCount` are derived from the Job specification. Admission happens later, and admission is anywhere between the two values.

Author: Sorry for confusion, I updated the section to be more accurate. 

The Workload's minCount must diverge from the original minCount for already admitted Jobs to ensure we don't admit fewer pods than are currently admitted.

Also, a new Workload representing the full job will be created and added to the queue, to admit the remaining capacity once it becomes available.

#### Opportunistic scale up when capacity is freed

> In order to schedule remaining pods after partial admission, workload controller will create a new workload representing the full job and add it to the queue. The scheduler will admit new workload and replace the old workload via workload slice mechanism as the capacity becomes available.

Reviewer 2: This sounds good. One detail is that we need to somehow generate the name for the workload probe that would not collide with the admitted workload. This is because currently ElasticJobs generate workload names based on the "base" of the name, and the Job's generation. However, both will be the same for the admitted workload, and the opportunistic probe. And I'm not totally sure how to solve it, maybe we could mix-in into the hash the "probe" type. 

I think we should at least flag the collision problem. I'm ok to say, that we propose using "opportunistic probe" to mix-into the hash, but details may be determined during implementation.

Author: Good point, updated.

#### Example: Two-Step Scale Up under Quota Constraints (with Partial Admission)

Consider a scenario where:
1. The ClusterQueue has a total quota of **7** for the requested resource flavor.
2. The Job is configured for both `ElasticJobs` and `PartialAdmission`.
3. The user performs a two-step scale up of the Job: starting at **5** replicas, scaling up to **10**, and then to **12**.

> ##### Step 0: Job Creation (Initial Size: 5)

Reviewer 2: Instead of "Step" please make there "Scenario" so that they can be analyzed independently

Author: >they can be analyzed independently

They are connected since the previous state was described in previous step (except step 0), so it make sense to process step 0 (previous state) + step 1 (current state).

> * **Job spec.parallelism**: 5

Reviewer 2: I think this example should also define for the Job the min-count annotation, otherwise it is unclear where `spec.podSets.minCount` is coming from

  3. **Kueue Scheduler**: Evaluates `wl-C`. The demand is `12 - 7 = 5`. The available quota is 0, so the scheduler puts `wl-C` in the queue.

##### Step 3: Quota increases to 12, opportunistic scale up when capacity is freed
If the available quota in the ClusterQueue increases to 12 (or more) in the future:
* **Workloads**:
  * `wl-B` (Finished)
  * `wl-C` (Admitted):
    * `spec.podSets.count` = 12
    * `spec.podSets.minCount` = 8 (7 + 1)
    * `status.admission.count` = 12

In the next scheduler loop, the Kueue scheduler will evaluate and admit `wl-C`. It will also update its `spec.podSets.minCount` to match the job's minCount.

### batch/Job controller

Besides adapting `RunWithPodSetsInfo` and `RestorePodSetsInfo` it should also:

- rework `PodSets()` to populate `MinCount` if the job is marked to support partial admission.
  * jobs supporting partial admission should have a dedicated annotation, e.g., `kueue.x-k8s.io/job-min-parallelism`, indicating the minimum `parallelism` acceptable by the job in case of partial admission.

>   * jobs which need the `completions` count kept in sync with `parallelism` should indicate this in a second annotation, `kueue.x-k8s.io/job-completions-equal-parallelism`

Reviewer 2: I think this will depend if we enable the elastic-jobs or not, right?
because it we enable the I assume workload's 
```
.spec.podSets.minCount=kueue.x-k8s.io/job-min-parallelism
.spec.podSets.count= Job's spec.parallelism
```

Author: >jobs which need the `completions` count kept in sync with `parallelism` should indicate this in a second annotation, `kueue.x-k8s.io/job-completions-equal-parallelism`

But this is unrelated to elastic-job. Not sure I understand your comment, can you clarify?

Whether an MPIJob supports partial admission or not can be deduced based on `MinAvailable` without the need of a dedicated annotation.
Additional research is needed into the potential usage of multiple variable count PodSets.

### RayJob/RayService/RayCluster controller

> The RayCluster.workerGroupSpec[i].minReplicas will be translated to the PodSet.MinCount for the Worker PodSet.

Reviewer 2: I think this is inaccurate., because we should probably use `minReplicas * numOfHosts`? 

Analogously as we do for count, ptal: https://github.com/kubernetes-sigs/kueue/blob/main/pkg/controller/jobs/raycluster/common.go#L69-L72

And please also describe the count so that the KEP also details for "count".

Author: Thanks, updated

  - `test/integration/singlecluster/controller/jobs/raycluster/raycluster_controller_test.go`:
    - `Should schedule RayClusters with partial admission order policy`: verifies a complete integration flow where a RayCluster with partial admission annotation is admitted with reduced worker count according to the order.
- **Workload Webhook**:
  - `test/integration/singlecluster/webhook/core/workload_test.go`:
    - `invalid podSet minCount (negative)`: verifies negative minCount values are rejected.
    - `invalid podSet minCount (too big)`: verifies minCount larger than count is rejected.
    - `too many variable count podSets`: verifies that workloads with multiple variable count PodSets are rejected.

#### E2E tests

- `test/e2e/singlecluster/baseline/job_test.go`:
  - `Should partially admit the Job if configured and not fully fits`: verifies that a real Job is successfully admitted with a reduced parallelism count matching the available cluster resources.

### Graduation Criteria

> 

Reviewer 2: Please fill in the gradution criteria for: PartialAdmissionForElasticJobs and PartialAdmissionForElasticJobsForBatchJob FGs, divide them in sections for Alpha, Beta and Stable.

Author: Added graduation criteria for PartialAdmissionForElasticJob

Whether an MPIJob supports partial admission or not can be deduced based on `MinAvailable` without the need of a dedicated annotation.

> Additional research is needed into the potential usage of multiple variable count PodSets.

Reviewer 2: I think we can update the KEP and drop this section. This is not done. Let's re-design when we do this

### RayJob/RayService/RayCluster controller

The RayCluster.workerGroupSpec[i].replicas * numOfHosts will be translated to the PodSet.Count, and minReplicas * numOfHosts will be translated to the PodSet.MinCount for the Worker PodSet.

### Limitations

Partial admission is not supported for concurrent admission to avoid increasing complexity. The webhook will reject the creation of workloads that support partial admission for a ClusterQueue with a concurrent admission policy. 

### Test Plan

No regressions in the current tests should be observed.

[X] I/we understand the owners of the involved components may require updates to
existing tests to make this code solid enough prior to committing the changes necessary
to implement this enhancement.

#### Unit Tests

> - **Scheduler/Flavor Assignment**:

Reviewer 2: I think the test plan needs refreshing, because it looks stale, for example the proporitonal mode is still here,

Author: removed proportional
